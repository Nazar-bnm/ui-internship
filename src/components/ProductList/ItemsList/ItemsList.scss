@import '../../../scss/global_styles';

.items-container {
  @include flex();
  width: 100%;

  &__error-container {
    width: 100%;
    padding: 30px 10px;
    text-align: center;
    color: red;
  }
}
