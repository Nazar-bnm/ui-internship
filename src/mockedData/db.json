{
  "employees": [
    {
      "id": 1,
      "first_name": "Sebastian",
      "last_name": "Eschweiler",
      "email": "sebastian@codingthesmartway.com"
    },
    {
      "id": 2,
      "first_name": "Steve",
      "last_name": "Palmer",
      "email": "steve@codingthesmartway.com"
    },
    {
      "id": 3,
      "first_name": "Ann",
      "last_name": "Smith",
      "email": "ann@codingthesmartway.com"
    }
  ],
  "what-is-new" : [
    {
      "id": 1,
      "photo": "https://www.next.ua/nxtcms/resource/image/1889302/portrait_ratio1x1/525/525/2fccc457347d46d0c506ee9ea0396ae8/of/emma-willis.jpg",
      "title": "Black and White",
      "collection": "spring collection 2016",
      "category": "women"
    },
    {
      "id": 2,
      "photo": "https://www.next.ua/nxtcms/resource/image/1233600/portrait_ratio1x1/525/525/69ca86774e073288bc792479650a14cc/Jn/all-baby-girl.jpg",
      "title": "Color Summer",
      "collection": "spring collection 2016",
      "category": "kids"
    },
    {
      "id": 3,
      "photo": "https://www.next.ua/nxtcms/resource/image/1915752/portrait_ratio1x1/525/525/1cf45576cf1de5e2bab187220d4ebeb0/ds/g83-mb-sh819338-lm-357.jpg",
      "title": "Vintage For Him",
      "collection": "spring collection 2016",
      "category": "men"
    }
  ],
  "most-popular": [
    {
      "id": 1,
      "image": "https://i.dailymail.co.uk/1s/2018/12/15/12/7458890-6498875-image-a-2_1544876240893.jpg",
      "label": "new",
      "title": "Black dress",
      "price": "99,50$",
      "sizes": "S - M - L"
    },
    {
      "id": 2,
      "image": "https://i.pinimg.com/564x/99/5d/f3/995df3032c828381332af1dea4e5c590.jpg",
      "label": "sale",
      "title": "Red evening dress",
      "price": "179.00$",
      "sizes": "S - M - L"
    },
    {
      "id": 3,
      "image": "https://i.redd.it/grla2qtpa5721.jpg",
      "label": "",
      "title": "Fit & flare dress",
      "price": "99.00$",
      "sizes": "S - M - L"
    }
  ]
}
