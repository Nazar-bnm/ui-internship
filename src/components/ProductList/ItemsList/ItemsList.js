import React from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';
import cx from 'classnames';

import ProductItem from '../../MostPopular/ProductItem';
import HttpService from '../../../service/HttpService/httpService';
import Preloader from '../../Preloader';

import mockedData from '../../../mockedDataForTests';
import { brandIDEnum } from '../../../constants';

import './ItemsList.scss';

const CN = 'items-container';
const userAPI = new HttpService();
const { mockedProductList } = mockedData;

class ItemsList extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      isLoading: true
    };

    this.sortByItemName = this.sortByItemName.bind(this);
    this.renderNoItemsMessage = this.renderNoItemsMessage.bind(this);
    this.getListItems = this.getListItems.bind(this);
  }

  componentDidMount() {
    this.getListItems();
  }

  componentDidUpdate(prevProps) {
    const { filters, itemsOnPage, currentPage } = this.props;

    if (!_.isEqual(prevProps.filters, filters)) {
      this.getListItems();
    }

    if (prevProps.itemsOnPage !== itemsOnPage || prevProps.currentPage !== currentPage) {
      this.getListItems();
    }
  }

  async getListItems() {
    const {
      categoryName,
      onGetProductsSuccess,
      onGetProductsError,
      gender,
      itemsOnPage,
      currentPage
    } = this.props;

    const categories = this.getCategoriesURL();
    const productsURL = `${process.env.SERVER_URL}/${categoryName}?genders=${gender}${categories}&limit=${itemsOnPage}&page=${currentPage}`;
    this.setState({
      isLoading: true
    });
    try {
      const response = await userAPI.get(productsURL);

      if (response.status === 404) {
        throw Error(response.statusText);
      }
      if (response.status === 500) {
        this.props.count = 0;
      }

      setTimeout(() => {
        this.setState({
          isLoading: false
        });
      }, 500);

      onGetProductsSuccess(response.data);
    } catch (error) {
      onGetProductsError(error);
    }
  }

  getItemsToRender() {
    const {
      error,
      wishlist,
      addToWishList,
      removeFromWishList,
      price
    } = this.props;

    const { fromPrice, toPrice } = price;

    if (error) {
      return (
        <div className={`${CN}__error-container`}>
          No items with selected parameters. Try running npm run mock:api to start the server and reload the page, or check the URL of your request.
        </div>
      );
    }

    const itemsToRender = this.sortItems().filter((item) => item.price > fromPrice && item.price < toPrice);

    return itemsToRender.map((el) => (
      <ProductItem
        key={el._id}
        product={el}
        wishlist={wishlist}
        addToWishList={addToWishList}
        removeFromWishList={removeFromWishList}
      />
    ));
  }

  getCategoriesURL() {
    const { filters } = this.props;
    const {
      bottoms,
      tops,
      sizes,
      colors,
      brands
    } = filters;
    let categories = '';
    const regexOnlyLetters = /[a-zA-Z]/g;
    const allFilters = bottoms.concat(tops, sizes, colors, brands);

    allFilters.forEach((item) => {
      if (bottoms.includes(item) || tops.includes(item)) {
        categories += `&category=${item}`;
      } else if (sizes.includes(item)) {
        const croppedItem = item.match(regexOnlyLetters).join('');
        categories += `&sizes=${croppedItem}`;
      } else if (colors.includes(item)) {
        categories += `&colors=${item}`;
      } else {
        item.replace(' ', '');
        categories += `&brandId=${brandIDEnum[item]}`;
      }
    });

    return categories;
  }

  sortByItemName = (prevEl, nextEl) => {
    const { ascendingOrder } = this.props;
    const prevName = prevEl.title.toUpperCase();
    const nextName = nextEl.title.toUpperCase();
    const isPrevLetterBigger = (prevName > nextName) ? 1 : 0;

    if (ascendingOrder) {
      return (prevName < nextName) ? -1 : isPrevLetterBigger;
    }
    return (prevName < nextName) ? isPrevLetterBigger : -1;
  }


  sortByPrice = (prevEl, nextEl) => {
    const { ascendingOrder } = this.props;

    return (ascendingOrder ? prevEl.price - nextEl.price : nextEl.price - prevEl.price);
  }

  sortItems() {
    const { ascendingOrder, orderType, itemList } = this.props;
    const itemsToRender = [...itemList];

    switch (orderType) {
      case 'Price':
        itemsToRender.sort((prevEl, nextEl) => this.sortByPrice(prevEl, nextEl));
        break;
      case 'Name':
        itemsToRender.sort((prevEl, nextEl) => this.sortByItemName(prevEl, nextEl));
        break;
      case 'Position':
        if (!ascendingOrder) {
          itemsToRender.reverse();
        }
        break;
      default:
        break;
    }

    return itemsToRender;
  }

  renderNoItemsMessage() {
    return (
      <div className={`${CN}__error-container`}>
        There are no products with selected parameters.
      </div>
    );
  }

  render() {
    const { isLoading } = this.state;

    return (
      <div className={cx(CN, { [`${CN}--loader`]: isLoading })}>{!isLoading ? this.getItemsToRender() : <Preloader />}</div>
    );
  }
}

ItemsList.propTypes = {
  currentPage: PropTypes.number,
  ascendingOrder: PropTypes.bool,
  categoryName: PropTypes.string,
  itemList: PropTypes.array,
  itemsOnPage: PropTypes.number,
  onGetProductsError: PropTypes.func,
  onGetProductsSuccess: PropTypes.func
};

ItemsList.defaultProps = {
  currentPage: 0,
  ascendingOrder: true,
  categoryName: 'products',
  itemList: mockedProductList,
  itemsOnPage: 6,
  onGetProductsError: null,
  onGetProductsSuccess: null
};

export default ItemsList;
