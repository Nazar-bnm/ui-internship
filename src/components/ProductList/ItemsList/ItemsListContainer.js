import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import ItemsList from './ItemsList';
import { getProductsSuccess, getProductsError } from '../../../actions/actionsProductList';
import { addToWishlist, removeFromWishlist } from '../../../actions/actionsWishlist';

const mapStateToProps = (state) => ({
  ascendingOrder: state.reducerProductList.ascendingOrder,
  error: state.reducerProductList.error,
  itemsOnPage: Number(state.reducerProductList.chosenItemsOnPage),
  itemList: state.reducerProductList.itemList,
  orderType: state.reducerProductList.orderType,
  wishlist: state.reducerWishlist.wishlist,
  filters: state.reducerFilter,
});
const mapDispatchToProps = (dispatch) => ({
  onGetProductsSuccess: bindActionCreators(getProductsSuccess, dispatch),
  onGetProductsError: bindActionCreators(getProductsError, dispatch),
  addToWishlist: bindActionCreators(addToWishlist, dispatch),
  removeFromWishlist: bindActionCreators(removeFromWishlist, dispatch)
});

export default connect(mapStateToProps, mapDispatchToProps)(ItemsList);
