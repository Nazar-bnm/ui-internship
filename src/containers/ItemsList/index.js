import { connect } from 'react-redux';
import ItemsList from '../../components/ProductList/ItemsList';
import { getProductsSuccess, getProductsError } from '../../actions/actionsProductList';

const mapStateToProps = (state) => {
  return {
    ascendingOrder: state.reducerProductList.ascendingOrder,
    orderType: state.reducerProductList.orderType,
    itemsOnPage: Number(state.reducerProductList.chosenItemsOnPage),
    itemList: state.reducerProductList.itemList,
    error: state.reducerProductList.error,
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    onGetProductsSuccess: (products) => dispatch(getProductsSuccess(products)),
    onGetProductsError: (error) => dispatch(getProductsError(error)),
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(ItemsList);
