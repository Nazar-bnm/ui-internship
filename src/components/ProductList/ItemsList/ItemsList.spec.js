import React from 'react';
import MockAdapter from 'axios-mock-adapter';
import axios from 'axios';
import { mount } from 'enzyme';

import ItemsList from './ItemsList';
import mockedData from '../../../mockedDataForTests';

describe('<ItemsList />', () => {
  let wrapper;
  const { mockedProductList } = mockedData;

  beforeEach(() => {
    const mock = new MockAdapter(axios);

    wrapper = mount(<ItemsList />);
    mock.onGet(`${process.env.BASE_URL}/product-list`).reply(200, mockedProductList);
  });

  test('should check `componentDidMount()`', () => {
    const instance = wrapper.instance();
    
    jest.spyOn(instance, 'getListItems');
    instance.componentDidMount();
    expect(instance.getListItems).toHaveBeenCalledTimes(1);
  });

  test('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
