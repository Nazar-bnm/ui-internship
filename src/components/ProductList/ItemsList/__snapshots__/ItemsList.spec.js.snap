// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ItemsList /> should match snapshot 1`] = `
<div
  className="items-container"
>
  <Connect(ProductItem)
    key="Awesome dress"
    product={
      Object {
        "category": "women",
        "collection": "spring collection 2016",
        "id": "1",
        "image": "https://img.ans-media.com/files/sc_staging_images/product/full_img_2033080.jpg",
        "label": "sale",
        "price": "199",
        "sizes": "[S, M, L]",
        "title": "Awesome dress",
      }
    }
  />
  <Connect(ProductItem)
    key="Beautiful dress"
    product={
      Object {
        "category": "women",
        "collection": "spring collection 2016",
        "id": "2",
        "image": "https://img.ans-media.com/files/sc_staging_images/product/large_img_2033076.jpg",
        "label": "new",
        "price": "299",
        "sizes": "[S, M, L]",
        "title": "Beautiful dress",
      }
    }
  />
  <Connect(ProductItem)
    key="Amazing dress"
    product={
      Object {
        "category": "women",
        "collection": "spring collection 2016",
        "id": "3",
        "image": "https://img.ans-media.com/files/sc_staging_images/product/large_img_2033787.jpg",
        "label": "",
        "price": "399",
        "sizes": "[M, L]",
        "title": "Amazing dress",
      }
    }
  />
</div>
`;
